flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/SessionDispatcherLeaderProcess.java

```diff
@@ -92,34 +99,61 @@ private void startServices() {
         }
     }
 
-    private void createDispatcherIfRunning(Collection<JobGraph> jobGraphs) {
-        runIfStateIs(State.RUNNING, () -> createDispatcher(jobGraphs));
+    private void createDispatcherIfRunning(
+            Collection<JobGraph> jobGraphs, Collection<JobResult> globallyTerminatedJobs) {
+        runIfStateIs(State.RUNNING, () -> createDispatcher(jobGraphs, globallyTerminatedJobs));
     }
 
-    private void createDispatcher(Collection<JobGraph> jobGraphs) {
+    private void createDispatcher(
+            Collection<JobGraph> jobGraphs, Collection<JobResult> globallyTerminatedJobs) {
 
         final DispatcherGatewayService dispatcherService =
                 dispatcherGatewayServiceFactory.create(
-                        DispatcherId.fromUuid(getLeaderSessionId()), jobGraphs, jobGraphStore);
+                        DispatcherId.fromUuid(getLeaderSessionId()),
+                        jobGraphs,
+                        globallyTerminatedJobs,
+                        jobGraphStore,
+                        jobResultStore);
 
         completeDispatcherSetup(dispatcherService);
     }
 
-    private CompletableFuture<Collection<JobGraph>> recoverJobsAsync() {
-        return CompletableFuture.supplyAsync(this::recoverJobsIfRunning, ioExecutor);
+    private CompletableFuture<Void>
+            createDispatcherBasedOnRecoveredJobGraphsAndGloballyTerminatedJobs() {
+        return CompletableFuture.supplyAsync(
+                        this::getGloballyCompletedJobResultsIfRunning, ioExecutor)
+                .thenCompose(
+                        globallyTerminatedJobs ->
+                                CompletableFuture.supplyAsync(
+                                                () ->
+                                                        this.recoverJobsIfRunning(
+                                                                globallyTerminatedJobs.stream()
+                                                                        .map(JobResult::getJobId)
+                                                                        .collect(
+                                                                                Collectors
+                                                                                        .toSet())),
+                                                ioExecutor)
+                                        .thenAccept(
+                                                jobGraphs ->
+                                                        createDispatcherIfRunning(
+                                                                jobGraphs, globallyTerminatedJobs))
+                                        .handle(this::onErrorIfRunning));
```

Comment:
Interesting, I didn't know about the `.thenAcceptBoth` method. That makes the code way easier to read 👍

Reply:
I'm wondering whether the `dirtyJobsFuture.thenApplyAsync` is necessary here. Wouldn't it be executed in the same thread pool as the previous `supplyAsync` since both are chained together?

Reply:
I've tried to reply via email, but it doesn't seem to have arrived :/

It would be executed in the same thread that has called the complete method (which belongs to the pool). Passing the executor directly allows another thread from the pool to execute the callback. Also the control mechanism is kind of more explicit as you can never be sure that the "complete" method has been called by the thread from previous stage.